packages/ruleset/src/functions/schema-naming-convention.js

```diff
@@ -0,0 +1,474 @@
+      // the paths, transforming them from Spectral's internal format, and maps
+      // them to the name of the schema they reference.
+      Object.keys(refMap).forEach(pathWithRef => {
+        const path = pathWithRef
+          .split('/')
+          .map(p => decodeURIComponent(p.replaceAll('~1', '/')))
+          .join('.')
+          .slice(2);
+        resultMap[path] = refMap[pathWithRef].slice(2).replaceAll('/', '.');
+      });
+    }
+  });
+
+  return resultMap;
+}
+
+/**
+ * Takes an unresolved path to a schema and un-resolves it to the format in which
+ * it will be stored in the graph nodes map by Spectral. The nodes provide a
+ * map from resolved path locations to the locations in 'components' they
+ * reference (if there is a reference at that location). Each path will be
+ * resolved to their nearest parent. For example if a request body schema
+ * references a schema in components, there will be an entry mapping the path
+ * to the request body schema to the schema path in components - but if that
+ * referenced schema has a property defined by a reference to another schema,
+ * there will not be an entry in the map including the path leading through
+ * the request body to the sub-property, there will be an entry mapping from
+ * the location of the schema in components. Here are example entries:
+ * - paths./v1/things.post.requestBody.content.application/json.schema:
+ *     components.schemas.ThingPrototype
+ * - components.schemas.ThingPrototype.properties.data:
+ *     components.schemas.DataObject
+ *
+ * The purpose of this function is to take the fully resolved path (in the
+ * example above, something like 'paths./v1/things.post.requestBody.content.
+ * application/json.schema.properties.data') and un-resolve enough that it
+ * will match the format Spectral uses in nodes (e.g. 'components.schemas.
+ * ThingPrototype.properties.data').
+ *
+ * @param {string} path - the resolved JSON path to a schema, as a dot-separated string
+ * @param {object} pathToReferencesMap - the graph nodes map from Spectral, which has
+ *                                       been sanitzed for our purposes. it maps referenced
+ *                                       schema locations to their reference location.
+ * @returns {string} - the name of the referenced schema at a given path, or undefined
+ */
+function getSchemaNameAtPath(path, pathToReferencesMap) {
+  if (!path && typeof path !== 'string') {
+    return;
+  }
+
+  // resolve path to reference - add comments explaining
+  // Build the path, replacing each path that resolves to a reference with the
+  // referenced path in order to match the expected format in the
+  // pathToReferencesMap (which comes from graph nodes that Spectral gives us).
+  // See the function documentation above for more info.
+  let pathBuilder = '';
+  for (const pathSegment of path.split('.')) {
+    pathBuilder += `${pathSegment}`;
```

Comment:
Another way to do this would be add a period to the string immediately before adding the next path segment (as long as pathBuilder already has SOMETHING in it):
```suggestion
    if (pathBuilder) {
      pathBuilder += '.';
    }
    pathBuilder += `${pathSegment}`;
```
then remove the `pathBuilder += '.';` line below

Reply:
that way you don't need to remove anything later on.

Reply:
Good call - I knew there was a better way to handle that but never ended up revisiting it. I'll make that change